# 개요
- 참고 자료를 보셔도 좋지만, 코드 이해뿐만 아니라 데이터 전처리 처리에 대한 순서 / 법칙을 이해하면서 작성합시다!
- 코드에 대한 다양한 도전을 합시다 :) 
- 원본 데이터 : https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset/data

## 문제 1: 데이터 불러오기와 크기 확인
뇌졸중 예측 데이터셋 CSV 파일을 pandas를 사용하여 불러오고, 데이터프레임의 크기(행과 열의 개수)를 출력하는 코드를 작성하세요. (힌트: pd.read_csv 함수를 사용합니다.)

In [16]:
import pandas as pd # 판다스 데이터조작

df = pd.read_csv('superstore_orders.csv')
df

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
1     51676  Female  61.0             0              0          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5105  18234  Female  80.0             1              0          Yes   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
1     Self-employed          Rural             202.21   NaN     never smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
...             ...            ...                ...   ...              ...   
5105        Private          Urban              83.75   NaN     never smoked   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
5105       0  
5106       0  
5107       0  
5108       0  
5109       0  

[5110 rows x 12 columns]

## 문제 2: 데이터 미리보기 (앞부분 확인)
데이터가 잘 불러와졌는지 확인하기 위해 데이터프레임의 처음 5행을 출력하는 코드를 작성하세요. 이때 데이터의 각 열 이름도 함께 확인해보세요. 그리고 각 열의 의미를 적어보세요!  
그리고 특정 열이 어떤 값이 있는지 확인해보세요! (특정열.unique() 사용)

In [7]:
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

## 문제 3: 데이터 정보 확인 (데이터 타입과 결측값)
데이터프레임의 각 열에 대한 요약 정보를 출력하는 코드를 작성하세요. 이를 통해 각 열의 데이터 타입(dtype)과 누락된 값(null/결측값) 개수를 확인해보세요. (힌트: DataFrame.info() 메서드를 사용합니다.)

In [ ]:
print(df.info(),"n/")



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB
None n/


(5110, 12)

## 문제 3-1: 데이터 정보 확인 (특정 컬럼 값 분포 확인)
hypertension, ever_married, avg_glucose_level, stroke 열의 값 분포를 확인하세요! (힌트: value_counts() 메서드를 사용합니다.)

In [15]:
df[['hypertension', 'ever_married', 'avg_glucose_level', 'stroke']].value_counts()

hypertension  ever_married  avg_glucose_level  stroke
0             Yes           91.68              0         5
                            114.32             0         4
                            72.49              0         4
              No            93.74              0         4
                            93.88              0         4
                                                        ..
              Yes           57.93              0         1
                            57.96              0         1
                            58.01              0         1
                            58.23              0         1
1             Yes           271.74             1         1
Name: count, Length: 4579, dtype: int64

## 문제 4: 결측값 개수 확인
각 열별로 결측값(null)이 몇 개인지 합계로 확인하려 합니다. 이를 계산하는 코드를 작성하세요. (힌트: isnull() 또는 isna()와 sum() 함수를 활용합니다.)

In [18]:
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

## 문제 5: id 열 삭제하기
id 열은 각 환자를 식별하기 위한 고유 식별자일 뿐, 예측 모델링에는 필요하지 않은 정보입니다. 데이터프레임에서 id 열을 제거하는 코드를 작성하세요.

## 문제 6: BMI 결측값 처리 (중앙값 대체)
bmi 열에 존재하는 201개의 결측값을 어떻게 처리할지 결정해야 합니다. 이번에는 중앙값(median)을 사용하여 이 결측치를 채우려고 합니다. bmi 열의 중앙값을 계산하고, 그 값으로 결측값들을 채우는 코드를 작성하세요.

## 문제 7: 행 삭제(drop) 연습
bmi 값이 결측치(NaN) 이거나, bmi < 10 또는 bmi > 80인 행은 잘못된 데이터라고 가정합시다.  
이런 행들을 제거한 새로운 데이터프레임 df_clean을 만드세요. (원본 df는 그대로 둡니다.)  
제거 전/후 행 개수를 비교해서 출력해보세요.  

## 문제 8: 조건 필터링 + 다중 정렬
나이 60세 이상이면서 (고혈압=1 또는 심장질환=1)인 행만 필터링해 생성  
avg_glucose_level 내림차순, 동률이면 bmi 오름차순으로 정렬  
상위 10명에 대해 id, age, avg_glucose_level, bmi, stroke만 출력  

## 문제 9: .loc vs .iloc 비교하기
데이터프레임 df에서 age, bmi, stroke 열만 뽑아서 확인해봅시다.
1. .loc를 사용해서 0~2번째 행의 age, bmi, stroke 값을 가져오세요.
2. .iloc를 사용해서 같은 행과 열을 가져오세요.
3. 두 결과가 같은지 눈으로 비교해보세요.